### PR TITLE
socket: add support to control no. of connections from one source

### PR DESCRIPTION
**Pull request by ssahani**
Introduce MaxConnectionsPerSource= that is number of concurrent
connections allowed per IP.

RFE: #1939

### CODE REVIEW

**Comment by martinpitt** [comment]
The boot-and-services and upstream QEMU tests seem to seriously break on this one. As it happened the same way on two architectures (these tests don't run on s390x) I'm marking this as failed.

**Comment by ssahani** [comment]
I readjusted the code to default to be disabled and it passed the CI.

File: man/systemd.socket.xml [hunk 1]
```diff
@@ -443,6 +443,14 @@
       </varlistentry>
 
       <varlistentry>
+        <term><varname>MaxConnectionsPerSource=</varname></term>
+        <listitem><para>The maximum number of connections from one source to
+        simultaneously run services instances for. It's works in conjuction with
```
Comment (fsateler) [inline man/systemd.socket.xml]: What does "in conjuction with" mean in this context? Is it that it is ignored if MaxConnections is not set?

Also, it should be "It", not "It's"

File: src/core/socket.c [hunk 2]
```diff
@@ -466,13 +482,57 @@ static int socket_verify(Socket *s) {
...
+                return 1;
+
+        switch(x->family) {
+        case AF_INET:
+                return memcmp(&x->peer.in.sin_addr, &y->peer.in.sin_addr, sizeof(x->peer.in.sin_addr));
+        case AF_INET6:
+                return memcmp(&x->peer.in6.sin6_addr, &y->peer.in6.sin6_addr, sizeof(x->peer.in6.sin6_addr));
+        }
+
+        return -1;
```
Comment (poettering) [inline src/core/socket.c:507]: you may replace the switch statement and final return with simply:

``` c
return memcmp(&x->peer, &y->peer, FAMILY_ADDRESS_SIZE(x->family));
```

File: src/core/socket.h [hunk 3]
```diff
@@ -162,11 +167,18 @@ struct Socket {
         RateLimit trigger_limit;
 };
 
+typedef struct SocketPeer {
+        unsigned n_connections;
+
+        int family;
+        union sockaddr_union peer;
+} SocketPeer;
```
Comment (poettering) [inline src/core/socket.h]: So, I you kinda already implement a reference counted object here, as each SocketPeer object is kept alive by the Service objects that share the same peer. I think we should make this explicit, and follow the usual logic for reference counted objects on this one too. Specifically, I think that "n_connections" should be renamed "n_ref" (since that's how we usually call the reference counter), and then each Service object of the same peer maintains a direct reference to it, i.e. gains a new `SocketPeer* peer` field directly in the Service structure.

That way the connections_by_address and connections_by_fd data structures in the Socket object can be replaced by a single "peers_by_address" hash table. And SocketPeer of course needs a ptr back to the Scoket.

I hope this makes sense?

File: src/core/socket.c (hunk 2 above)
Comment (ssahani) [inline src/core/socket.c:507]: I tried with it seems not working.

File: src/core/socket.h (hunk 3 above)
Comment (ssahani) [inline src/core/socket.h]: Yes I changed it now . I was thinking about putting the peer pointer but then din't wanted to add a extra variable for Service structure.

File: src/core/socket.h [hunk 4]
```diff
@@ -162,11 +167,18 @@ struct Socket {
         RateLimit trigger_limit;
 };
 
+struct SocketPeer {
+        unsigned n_ref;
+
+        int family;
+        union sockaddr_union peer;
+};
```
Comment (poettering) [inline src/core/socket.h:176]: this really should have a ptr back to the Socket object. (see above)

File: src/core/socket.c [hunk 5]
```diff
@@ -2816,6 +2927,19 @@ void socket_connection_unref(Socket *s) {
         assert(s->n_connections > 0);
         s->n_connections--;
 
+        if (service->peer) {
+                service->peer->n_ref--;
```
Comment (poettering) [inline src/core/socket.c]: please move this into the usual service_peer_unref() call, as we typically (though not always) do for reference counted objects.

File: src/core/socket.c [hunk 6]
```diff
@@ -141,11 +143,20 @@ void socket_free_ports(Socket *s) {
...
+        Iterator i;
 
         assert(s);
 
         socket_free_ports(s);
 
+        HASHMAP_FOREACH(p, s->peers_by_address, i) {
+                hashmap_remove(s->peers_by_address, p);
+                free(p);
+        }
```
Comment (poettering) [inline src/core/socket.c]: this will explode, no? I mean what happens if a Service object referencing one of these SocketPeer objects is cleaned up after the Socket object? in that case its reference to the SocketPeer will point to an object that is already freed.

I think it would be best to allow SocketPeer to exist either attached to a Socket, or detached from it. The latter would happen only if the Socket it originally belonged to got destroyed, but the SocketPeer is still referenced by a Socket. I'd encode the fact whether a SocketPeer object is attached or detached simply by including a ptr from the SocketPeer to its Socket object, which would be NULL if the object is detached, and non-NULL if still attached.

File: src/core/socket.c (hunk 2 above)
Comment (poettering) [inline src/core/socket.c:507]: any idea why?

File: src/core/socket.c [hunk 7]
```diff
@@ -2087,6 +2154,11 @@ static void socket_enter_running(Socket *s, int cfd) {
                 cfd = -1; /* We passed ownership of the fd to the service now. Forget it here. */
                 s->n_connections++;
 
+                if (s->max_connections_per_source > 0 ) {
+                        p->n_ref++;
+                        service->peer = p;
+                }
```
Comment (poettering) [inline src/core/socket.c:2169]: please add the usual socket_peer_ref() call here too!

File: src/core/socket.c [hunk 8]
```diff
@@ -2542,6 +2614,45 @@ _pure_ static bool socket_check_gc(Unit *u) {
         return s->n_connections > 0;
 }
 
+int socket_find_connection(Socket *s, int fd, SocketPeer **p) {
```
Comment (poettering) [inline src/core/socket.c]: this should be called socket_find_peer(), no?

File: src/core/service.c [hunk 9]
```diff
@@ -188,7 +188,7 @@ void service_close_socket_fd(Service *s) {
         s->socket_fd = asynchronous_close(s->socket_fd);
 
         if (UNIT_ISSET(s->accept_socket)) {
-                socket_connection_unref(SOCKET(UNIT_DEREF(s->accept_socket)));
+                socket_connection_unref(SOCKET(UNIT_DEREF(s->accept_socket)), s);
                 unit_ref_unset(&s->accept_socket);
         }
```
Comment (poettering) [inline src/core/service.c:193]: when the service is freed it should really release its "peer" refernce should it still have one. i.e. there should be socket_peer_unref() and it should be called from service_done() on the service's peer object.

File: src/core/socket.c (hunk 6 above)
Comment (ssahani) [inline src/core/socket.c]: Yes  now I moved the clean up of the peer when the service gets stop itself. That is we remove from hashmap and free the peer memory. But we still need to clean up the hashmap memory itself.  I think there will be a memory leak . Inside the peer object if we store the hashmap pointer then we can access it after Socket object gets freed also. 
1. socket starts running let's say 3 telnet clients. Here we store the peer address that contains
   the socket address.

```
struct SocketPeer {
        unsigned n_ref;
        Socket *s;
        int family;
        union sockaddr_union peer;
};
```
1. If we restart the socket the service object has invalid memory which is actually freed that is the pointer 's' . So we can only decrement the counter but never able remove from hash table since it's freed. And it will be there forever. How to deal with this ? I am pushing the code which is crashing when we try to remove from hashmap.

Comment (ssahani) [inline src/core/socket.c]: Sorry I think I misinterpreted . We can always remove from hash map and free.

File: src/core/socket.c [hunk 10]
```diff
@@ -2039,13 +2089,29 @@ static void socket_enter_running(Socket *s, int cfd) {
...
+                if (s->max_connections_per_source > 0) {
+
+                        r = socket_find_peer(s, cfd, &p);
+                        if (r < 0) {
+                                safe_close(cfd);
+                                return;
+                        }
+
+                        if (p->n_ref > s->max_connections_per_source) {
+                                log_unit_warning(UNIT(s), "Too many incoming connections (%u) from source, refusing connection attempt.", p->n_ref);
```
Comment (keszybz) [inline src/core/socket.c:2118]: Shouldn't this be ratelimited? It is under external control, so if the other party know a service has this set they can easily saturate logs with minimal effort.

The same applies to `s->n_connections >= s->max_connections` above. Adding ratelimiting to both might be a good idea for another PR.

File: src/core/socket.c (hunk 2 above)
Comment (keszybz) [inline src/core/socket.c:507]: What does it mean "not working"? Please provide some details.

File: src/core/socket.c [hunk 11]
```diff
@@ -141,11 +144,18 @@ void socket_free_ports(Socket *s) {
...
         Socket *s = SOCKET(u);
+        SocketPeer *p;
+        Iterator i;
 
         assert(s);
 
         socket_free_ports(s);
 
+        HASHMAP_FOREACH(p, s->peers_by_address, i)
+                hashmap_remove(s->peers_by_address, p);
```
Comment (poettering) [inline src/core/socket.c]: It's important to somehow encode in the SocketPeer object whether it is currently included in a Socket unit's "peers_by_address" hashmap or not. A natural way to encode this would be by using the "socket" ptr that is already part of SocketPeer. Simply let it point ot the right Socket\* structure as long as it's Socket is still around. Reset it to NULL as soon as its Socket is gone. Or in other words, rewrite the loop above to something like this:

``` c
while ((p = hashmap_steal_first(i)))
        p->socket = NULL;
```

File: src/core/socket.c (hunk 2 above)
Comment (poettering) [inline src/core/socket.c:507]: ah, actually @ssahani is right, this is a "union socket_addr_union", not a "union in_addr_union", hence doing my memcmp() suggestion doesn't really work.

That said, @ssahani I think it would make sense to rework this so that SocketPeer only stores an "union in_addr_union" instead of a full socket address, after all we don't know what to do with non-IP addresses anyway, and don#t care for port numbers either.

File: src/core/socket.c [hunk 12]
```diff
@@ -2039,13 +2089,29 @@ static void socket_enter_running(Socket *s, int cfd) {
...
+                                safe_close(cfd);
+                                return;
+                        }
+
+                        if (p->n_ref > s->max_connections_per_source) {
+                                log_unit_warning(UNIT(s), "Too many incoming connections (%u) from source, refusing connection attempt.", p->n_ref);
+                                safe_close(cfd);
+                                return;
+                        }
+                }
```
Comment (poettering) [inline src/core/socket.c:2123]: memory management appears borked here. You need to make sure that the socket peer object is automatically released when something fails here, if it was just created for this. My recommendation would be to make sure that socket_find_peer() always increases the ref cntr of the returned object by one on success, and the code here uses `_cleanup_(socket_peer_unrefp)` when declaring the SocketPeer, to ensure it is released again if something fails here...

File: src/core/socket.c [hunk 13]
```diff
@@ -2087,6 +2153,11 @@ static void socket_enter_running(Socket *s, int cfd) {
                 cfd = -1; /* We passed ownership of the fd to the service now. Forget it here. */
                 s->n_connections++;
 
+                if (s->max_connections_per_source > 0 ) {
+                        socket_peer_ref(p);
+                        service->peer = p;
```
Comment (poettering) [inline src/core/socket.c]: this is prettier when written like this:

``` c
service->peer = socket_peer_ref(p);
```

In particular, as you can then drop the {}...

File: src/core/socket.c [hunk 14]
```diff
@@ -2542,6 +2613,77 @@ _pure_ static bool socket_check_gc(Unit *u) {
         return s->n_connections > 0;
 }
 
+SocketPeer *socket_peer_new(void) {
+        SocketPeer *p;
+
+        p = new0(SocketPeer, 1);
+        if (!p)
+                return NULL;
```
Comment (poettering) [inline src/core/socket.c:2670]: shouldn't n_ref be initialized to 1 here?

File: src/core/socket.c [hunk 15]
```diff
@@ -2542,6 +2613,77 @@ _pure_ static bool socket_check_gc(Unit *u) {
...
+SocketPeer *socket_peer_unref(SocketPeer *p) {
+        if (!p)
+                return NULL;
+
+        assert(p->n_ref > 0);
+
+        p->n_ref--;
+
+        if (p->n_ref == 0)
+                free(p);
```
Comment (poettering) [inline src/core/socket.c]: if p->socket is set you should drop the SocketPeer object from p->socket->socket_peers really.

File: src/core/socket.c [hunk 16]
```diff
@@ -2542,6 +2613,77 @@ _pure_ static bool socket_check_gc(Unit *u) {
...
+
+        remote = socket_peer_new();
+        if (!remote)
+                return log_oom();
+
+        l = sizeof(remote->peer.sa);
+        r = getpeername(fd, (struct sockaddr*) &remote->peer.sa, &l);
+        if (r < 0) {
+                log_error_errno(errno, "getpeername failed: %m");
+                return -errno;
```
Comment (poettering) [inline src/core/socket.c]: log_error_errno() returns the negative error number you passed in, anyway. You can hence simplify these two lines to:

``` c
return log_error_errno(errno, "…");
```

and get rid of of the {}...

File: src/core/socket.c [hunk 17]
```diff
@@ -2542,6 +2613,77 @@ _pure_ static bool socket_check_gc(Unit *u) {
...
+
+        i = hashmap_get(s->peers_by_address, remote);
+        if (i) {
+                *p = i;
+                return 1;
+        }
+
+        r = hashmap_put(s->peers_by_address, remote, remote);
+        if (r < 0)
+                return r;
```
Comment (poettering) [inline src/core/socket.c:2734]: for every call of socket_find_peer() you do a memory allocation right now, and quite often for no reason, as we might return an existing object after all. 

I'd propose to change the hashmap to not map a SocketPeer onto a SocketPeer but instead map a "union in_addr_union" onto a SocketPair. That makes it very natural not to have to do allocate memory to search for a SocketPair, as you can pass the in_addr_union structure directly to the hashmap_get(), and only allocate a new object if that fails.

**Comment by poettering** [comment]
I did another round of reviews. Sorry for this taking so long!

The socket peer logic needs some kind of handling to deal with PID 1 reexecution. The SocketPair object needs to be reaacquired on deserialization of a service.

File: src/core/socket.c (hunk 17 above)
Comment (ssahani) [inline src/core/socket.c:2734]: I guess I missed something here.  `in_addr_union` does not have the family which is needed  the hash function to work . we can use existing `in_addr_data' .  So the other way if I follow you correctly we do a`getpeername`using`union sockaddr_union`local variable , extract the relevant data and copy to a local variable and feed this to hash. If we find we return the pointer of the`SocketPeer`object else we allocate a new`SocketPeer`  and copy again ?

File: src/core/socket.c (hunk 2 above)
Comment (ssahani) [inline src/core/socket.c:507]: Thanks I could not catch that . But I think we need the family too.

File: src/core/socket.c (hunk 10 above)
Comment (ssahani) [inline src/core/socket.c:2118]: Yes.

File: src/core/socket.c (hunk 11 above)
Comment (ssahani) [inline src/core/socket.c]: Ok

File: src/core/socket.c (hunk 12 above)
Comment (ssahani) [inline src/core/socket.c:2123]: yes thanks changed now.

File: src/core/socket.c (hunk 17 above)
Comment (poettering) [inline src/core/socket.c:2734]: Oh. You are right. We need the family hence sockaddr_union is the right choice. Sorry for the confusion!

**Comment by ssahani** [comment]
About the PID 1 re execution  same as we restart the `.socket` . We loose the information.
Should we serialize  the whole hashtable data  ?

File: src/core/socket.h [hunk 18]
```diff
@@ -164,6 +168,21 @@ struct Socket {
         RateLimit trigger_limit;
 };
 
+struct SocketPeer {
+        unsigned n_ref;
+
+        Socket *socket;
+
+        int family;
+        union sockaddr_union peer;
```
Comment (poettering) [inline src/core/socket.h:175]: given that we stick to sockaddr_union now, the separate "family" field is redundant, no? It's available as peer.sa.sa_family after all, too

File: src/core/socket.h [hunk 19]
```diff
@@ -164,6 +168,21 @@ struct Socket {
...
+};
+
+SocketPeer *socket_peer_ref(SocketPeer *p);
+SocketPeer *socket_peer_unref(SocketPeer *p);
+
+DEFINE_TRIVIAL_CLEANUP_FUNC(SocketPeer*, socket_peer_unref);
+#define _cleanup_link_unref_ _cleanup_(socket_peer_unrefp)
```
Comment (poettering) [inline src/core/socket.h]: we try to avoid such type specific cleanup macro definitions these days, except for the most common of types. Hence, please use use _cleanup_(socket_peer_unrefp) everywhere, and avoid defining the _cleanup_link_unref_ macro as a shortcut for that.

File: src/core/socket.c [hunk 20]
```diff
@@ -468,13 +477,53 @@ static int socket_verify(Socket *s) {
...
+
+        assert(s);
+
+        siphash24_compress(&s->family, sizeof(s->family), state);
+
+        if (s->family == AF_INET)
+                siphash24_compress(&s->peer.in.sin_addr, sizeof(s->peer.in.sin_addr), state);
+        else if (s->family == AF_INET6)
+                siphash24_compress(&s->peer.in6.sin6_addr, sizeof(s->peer.in6.sin6_addr), state);
+}
```
Comment (poettering) [inline src/core/socket.c:490]: The hashfunc/comparison func really should just operate on on the sockaddr_union and not on the SocketPeer at all. The sockaddr contains everything we need after all: the address family as well as the actual address.

**Comment by poettering** [comment]
> About the PID 1 re execution same as we restart the .socket . We loose the information.
> Should we serialize the whole hashtable data ?

THis isn't really necessary, as long as the table is correctly rebuild when we deserialize state.

File: src/core/socket.h (hunk 18 above)
Comment (ssahani) [inline src/core/socket.h:175]: Actually changed now.

File: src/core/socket.h (hunk 19 above)
Comment (ssahani) [inline src/core/socket.h]: Done.

File: src/core/socket.c (hunk 20 above)
Comment (ssahani) [inline src/core/socket.c:490]: OK

File: man/systemd.socket.xml [hunk 21]
```diff
@@ -443,6 +443,14 @@
       </varlistentry>
 
       <varlistentry>
+        <term><varname>MaxConnectionsPerSource=</varname></term>
+        <listitem><para>The maximum number of connections for a service per source IP address.
+        This is is very similar to the <varname>MaxConnections=</varname> directive
+        above. Disabled by default.</para>
```
Comment (keszybz) [inline man/systemd.socket.xml:449]: Would be nicer to say "Unlimited by default".

Comment (ssahani) [inline man/systemd.socket.xml:449]: I think eventually this will be hit by `MaxConnections=` that is 64 .

File: src/core/load-fragment-gperf.gperf.m4 [hunk 22]
```diff
@@ -293,6 +293,7 @@ Socket.DirectoryMode,            config_parse_mode,                  0,
 Socket.Accept,                   config_parse_bool,                  0,                             offsetof(Socket, accept)
 Socket.Writable,                 config_parse_bool,                  0,                             offsetof(Socket, writable)
 Socket.MaxConnections,           config_parse_unsigned,              0,                             offsetof(Socket, max_connections)
+Socket.MaxConnectionsPerSource,  config_parse_unsigned,              0,                             offsetof(Socket, max_connections_per_source)
```
Comment (keszybz) [inline src/core/load-fragment-gperf.gperf.m4:296]: This should also be exported in `bus_socket_vtable`, similarly to MaxConnections. Otherwise systmectl show -p MaxConnectionsPerSource does not show this property.

File: src/core/socket.c (hunk 2 above)
Comment (keszybz) [inline src/core/socket.c:507]: You can store `union in_addr_union` and the family as two fields.

File: src/core/socket.h [hunk 23]
```diff
@@ -79,9 +80,12 @@ struct Socket {
 
         LIST_HEAD(SocketPort, ports);
 
+        Hashmap *peers_by_address;
```
Comment (keszybz) [inline src/core/socket.h:83]: This should be a set instead.

**Comment by keszybz** [comment]
Something seems wrong with the counting of the number of connections. Just opening new connections indeed blocks at the specified limit, but if I forcibly stop a connection with 'systemctl stop testlimit@...service', I'm able to start two new connections afterwards. Not sure what is going on here, but it doesn't seem right.

**Comment by ssahani** [comment]
@keszybz Please elaborate the test case. Let me try the same .

File: src/core/load-fragment-gperf.gperf.m4 (hunk 22 above)
Comment (ssahani) [inline src/core/load-fragment-gperf.gperf.m4:296]: yes

**Comment by keszybz** [comment]
```
# service.py
import time
import sys
import os
from systemd import journal

journal.send("Connection from {}:{}".format(os.getenv('REMOTE_ADDR'), os.getenv('REMOTE_PORT')))

time.sleep(1000)

# /etc/systemd/system/testlimit2.socket
[Socket]
MaxConnectionsPerSource=3
ListenStream=2000
Accept=yes

# /etc/systemd/system/testlimit2@.service
[Service]
ExecStart=/usr/bin/python3 /var/tmp/service.py

User=hasher
NoNewPrivileges=yes
ProtectHome=yes
ProtectSystem=full

PrivateNetwork=yes
```

Then connect from a second computer until I hit the limit:

```
telnet 10.48.19.205 2000 &
telnet 10.48.19.205 2000 &
telnet 10.48.19.205 2000 &
telnet 10.48.19.205 2000 &
```

And stop some of the services with `sudo systemctl stop testlimit2@20-10.48.19.205:2000-….service`
and repeat the connections. At this point `system list-units testlimit2@*` lists more than three services.

**Comment by ssahani** [comment]
fixed now !

File: src/core/socket.c [hunk 24]
```diff
@@ -2554,6 +2662,83 @@ _pure_ static bool socket_check_gc(Unit *u) {
...
+        remote->socket = s;
+
+        r = hashmap_put(s->peers_by_address, remote, remote);
+        if (r < 0)
+                return r;
+
+        *p = remote;
+        remote = NULL;
+
+        return 0;
```
Comment (poettering) [inline src/core/socket.c:2739]: This call really doesn't look right. If the peer object doesn't exist yet it will be created, the ref count will be set to 1 and that is returned. If the peer object exists already, it will be returned as is, the refcount is not changed. This means that a caller which calls socket_peer_unref() after acquiring the object with this function will destroy the object if it was just created (which is OK), but if it already exists, it will take away an earlier reference! And that's not OK.

I am pretty sure this funciton should increase the ref counter by one if the object already exists.

File: src/core/socket.c [hunk 25]
```diff
@@ -2058,6 +2107,21 @@ static void socket_enter_running(Socket *s, int cfd) {
                         return;
                 }
 
+                if (s->max_connections_per_source > 0) {
```
Comment (poettering) [inline src/core/socket.c:2110]: our hash op functions for the hashmap cannot deal with address family that aren't AF_INET or AF_INET6. That's OK, but this means we should refuse the counting here for other address family. Otherwise we'll end up doing the tracking half-way for AF_UNIX for example, where all this would semi-break.

File: src/core/socket.c [hunk 26]
```diff
@@ -2295,6 +2366,16 @@ static int socket_serialize(Unit *u, FILE *f, FDSet *fds) {
...
 
+        HASHMAP_FOREACH(k, s->peers_by_address, i) {
+                _cleanup_free_ char *t = NULL;
+
+                r = sockaddr_pretty(&k->peer.sa, FAMILY_ADDRESS_SIZE(k->peer.sa.sa_family), true, true, &t);
+                if (r < 0)
+                        return r;
+
+                unit_serialize_item_format(u, f, "peer", "%u %s", k->n_ref, t);
+        }
```
Comment (poettering) [inline src/core/socket.c:2377]: I don't think we need to serialize anything here tbh. The information you are serializing here can totally be reconstructed during deserialization directly from the socket connection anyway. Just call socket_find_peer() again for the connection fds and the peer table should be reconstructued without and further need for serialization.

File: src/core/socket.c [hunk 27]
```diff
@@ -2554,6 +2662,83 @@ _pure_ static bool socket_check_gc(Unit *u) {
...
+
+        if (p->socket)
+                (void) hashmap_remove(p->socket->peers_by_address, p);
+
+        free(p);
+
+        return NULL;
+}
+
+int socket_find_peer(Socket *s, int fd, SocketPeer **p) {
```
Comment (poettering) [inline src/core/socket.c:2706]: i think socket_acquire_peer() would be a better name for this. "find" has a very "passive" feeling I think, i.e. the name suggests that the call simply looks for an existing peer. But this does more actually, it actively allocates a new peer object if it doesn't exist.

File: src/core/socket.c [hunk 28]
```diff
@@ -2058,6 +2107,21 @@ static void socket_enter_running(Socket *s, int cfd) {
...
+                        if (r < 0) {
+                                safe_close(cfd);
+                                return;
+                        }
+
+                        if (p->n_ref > s->max_connections_per_source) {
+                                log_unit_warning(UNIT(s), "Too many incoming connections (%u) from source, refusing connection attempt.", p->n_ref);
+                                safe_close(cfd);
+                                p = NULL;
+                                return;
```
Comment (poettering) [inline src/core/socket.c:2121]: this looks strange. you set p = NULL here, but what if the peer object was just allocated? in that case you leak the object here. Also see my comments on socket_find_peer() below.

**Comment by poettering** [comment]
Hmm, I did another review, post-merge. There are some serious issues left regarding ref counting. This really needs more work.

**Comment by poettering** [comment]
I now filed issue #3878, and marked it as blocker for v232, since this really needs fixing before the release and we shouldn't forget it.
